draft-ietf-tls-tls13.md

```diff
@@ -2810,32 +2827,82 @@ configuration identifier for each connection, thus preventing
 replay. Implementations are responsible for ensuring uniqueness of the
 identifier in this case.
 
+The "ticket_age" extension sent by the client SHOULD be used by
+servers to limit the time over which the first flight might be
+replayed.  A server can store the time at which it sends a server
+configuration to a client, or encode the time in a ticket.  Then, each
+time it receives an early_data extension, it can check to see if the
+value used by the client matches its expectations.
+
+The "ticket_age" value provided by the client will be shorter than the
+actual time elapsed on the server by a single round trip time.  This
+difference is comprised of the delay in sending the NewSessionTicket
+message to the client, plus the time taken to send the ClientHello to
+the server.  For this reason, a server SHOULD measure the round trip
+time prior to sending the NewSessionTicket message and account for
+that in the value it saves.
+
+There are several potential sources of error that make an exact
+measurement of time difficult.  Variations in client and server clocks
+are likely to be minimal, outside of gross time corrections.  Network
+propagation delays are most likely causes of a mismatch in legitimate
+values for elapsed time.  Both the NewSessionTicket and ClientHello
+messages might be retransmitted and therefore delayed, which might be
+hidden by TCP.
+
+A small allowance for errors in clocks and variations in measurements
+is advisable.  However, any allowance also increases the opportunity
+for replay.  In this case, it is better to reject early data than to
+risk greater exposure to replay attacks.
+
+#### Ticket Age
+
+%%% Key Exchange Messages
+       struct {
+           uint64 ticket_age;
+       } TicketAge;
+
+When the client sends the "early_data" extension, it MUST also send
+a "ticket_age" extension in its EncryptedExtensions block. This value
+contains the time elapsed since the client learned about the server
+configuration that it is using, in milliseconds.  This value can
+be used by the server to limit the time over which early data can
+be replayed.
```

Comment:
Me too, but this is your text. Recommendations wanted

On Tue, May 3, 2016 at 10:07 PM, Dave Garrett <notifications@github.com>
wrote:

> In draft-ietf-tls-tls13.md
> <https://github.com/tlswg/tls13-spec/pull/449#discussion_r61990780>:
>
> > +for replay.  In this case, it is better to reject early data than to
> > +risk greater exposure to replay attacks.
> > +
> > +#### Ticket Age
> > +
> > +%%% Key Exchange Messages
> > +       struct {
> > +           uint64 ticket_age;
> > +       } TicketAge;
> > +
> > +When the client sends the "early_data" extension, it MUST also send
> > +a "ticket_age" extension in its EncryptedExtensions block. This value
> > +contains the time elapsed since the client learned about the server
> > +configuration that it is using, in milliseconds.  This value can
> > +be used by the server to limit the time over which early data can
> > +be replayed.
>
> I would really prefer if some recommended time window(s) were specified,
> rather than leaving this with no default sense of scale. A fairly large
> hard limit, in addition to more reasonable shorter recommendations, could
> also be good. I don't know what an ideal set of recommendations would be
> here, but leaving it unbounded doesn't feel great. Likewise, a general
> definition of what constitutes a "small allowance for errors" is would be
> helpful.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/tlswg/tls13-spec/pull/449/files/789724d7b33c4aa8a514c238abd2fca939240ea2..121d545b5235d320f66ee56de54efe85c7218de2#r61990780>
>